src/Kestrel.Core/Internal/Http/Http1MessageBody.cs

```diff
@@ -236,7 +73,9 @@ private void TryStopTimingReads()
                     BadHttpRequestException.Throw(RequestRejectionReason.UpgradeRequestCannotHavePayload);
                 }
 
-                return new ForUpgrade(context);
+                var body = new ForUpgrade(context);
+                context.RequestBodyPipeReader = new Http1MessageBodyPipeReader(context, body);
```

Comment:
This allocates a new PipeReader per request. Ideally we could just reuse this like we reuse the Stream. It should be resettable.